android-bindings/src/bindings.rs

```diff
@@ -1777,8 +1780,9 @@ pub unsafe extern "C" fn Java_com_mobilecoin_lib_TransactionBuilder_add_1output(
             let recipient: MutexGuard<PublicAddress> =
                 env.get_rust_field(recipient, RUST_OBJ_FIELD)?;
 
-            let mut rng = McRng::default();
-            let tx_out_context = tx_builder.add_output(value as u64, &recipient, &mut rng)?;
+            let mut rng: MutexGuard<ChaCha20Rng> =
```

Comment:
IIUC this adds support for using `McRng` or `ChaCha20Rng`, but we unconditionally use `ChaCha20Rng` in this body, which doesn't seem safe

Reply:
I think @remoun is right, why is `McRng` being exposed (as `DefaultRng`) below?

Reply:
> IIUC this adds support for using `McRng` or `ChaCha20Rng`, but we unconditionally use `ChaCha20Rng` in this body, which doesn't seem safe

Yes, this PR does expose both McRng and ChaCha20Rng to the SDK. However, you can only pass ChaCha20Rng to TransactionBuilder from the Java code. I'm not sure if there is a good, safe way to have interfaces/traits work across the native interface in the way you're saying, so I just decided to use ChaCha20Rng for TransactionBuilder always.

Reply:
> I think @remoun is right, why is `McRng` being exposed (as `DefaultRng`) below?

DefaultRng is just the name of the Java class that wraps McRng::default(). If you really think it should be renamed in the SDK to McRng or something, I could do that. But that name is just used below so it links to the correct class in Java.

Reply:
There's no way to wrap interfaces/traits. You either need to expose multiple variants, or you create a wrapper enum type and then expose that (for example `AttestedClientState` in this file).
When would `DefaultRng`/`McRng` get used?

Reply:
see the link above. It is exposed to the public API for use as a non-seedable RNG. It is also used internally to create a ChaCha20Rng that is seeded with a random value. This can be useful in the case where I am creating a transaction for the first time and don't care what the seed is—only that I can use it again later.
```Java
ChaCha20Rng rng = ChaCha20Rng.withRandomSeed();
byte seed[] = rng.getSeed();
// ... save seed and then later
// ...
ChaCha20Rng rng = ChaCha20Rng.fromSeed(seed);
// build same transaction again
```

Reply:
I see. Wouldn't be easier to add a method that creates a ChaCha20Rng with a randomly generated seed in the Rust side instead of exposing DefaultRng and potentially having someone accidentally pass a DefaultRng where ChaCha20Rng is expected?

Reply:
It isn't possible to pass a DefaultRng here. The functions here look like they take a generic Java Object, but they link to methods on the Java side that are type checked at compile time to be a ChaCha20Rng. But I wanted to expose the idea of an RNG to the SDK because it has a lot of other potential applications. There isn't a strict need to include it, but it serves as a default implementation of the Rng interface. ChaCha20Rng implements a SeedableRng interface which extends that. But it allows SDK users to get an RNG implementation for other purposes without dealing with the added complexity

Reply:
In principle the type can be some `Arc<dyn CryptoRngCore>` type that is bound by Java? I think we are doing that also somewhere else, maybe for memo builders

Reply:
> In principle the type can be some `Arc<dyn CryptoRngCore>` type that is bound by Java?

This was my initial goal, but I didn't find a way to make it work. I'll take a look at the memo builders and see if I can figure something out. It would be great if we could accept any suitable RNG here. On the Java side though, instead of just passing an RNG to TransactionBuilder, we would need to pass some final class that supplies an RNG to the bindings. Otherwise SDK users might assume they can implement the Rng interface and pass a custom RNG through.

While that makes usage a little more complex, I think it is worth it if we can easily change which RNGs are available without breaking the public API.